ykinfo.1
.\" Copyright (c) 2011-2014 Yubico AB
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above
.\"       copyright notice, this list of conditions and the following
.\"       disclaimer in the documentation and/or other materials provided
.\"       with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" The following commands are required for all man pages.
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.TH ykinfo "1" "October 2012" "yubikey-personalization"
.SH NAME
ykinfo - Get basic information from a YubiKey
.SH SYNOPSIS
.B ykinfo
[\fI-s\fR] [\fI-m\fR] [\fI-H\fR] [\fI-v\fR] [\fI-t\fR] [\fI-p\fR] [\fI-1\fR] [\fI-2\fR] [\fI-q\fR] [\fI-V\fR] [\fI-h\fR]
.SH DESCRIPTION
.PP
Get basic information from a YubiKey. Can get serial number, version
and current touch level. Can get all parameters at once or one at a
time.  If \fB\-q\fR is given only the relevant string from the YubiKey
is written to screen.
.SH OPTIONS
.TP
\fB\-s\fR
get the serial number and output it.
.TP
\fB\-m\fR
get the serial number and output it as modhex.
.TP
\fB\-H\fR
get the serial number and output it as hex.
.TP
\fB\-v\fR
get the version and output it.
.TP
\fB\-t\fR
get the touch level and output it.
.TP
\fB\-p\fR
get the programming sequence and output it.
.TP
\fB\-1\fR
get the status of slot 1, 1 for programmed and 0 for empty.
.TP
\fB\-2\fR
get the status of slot 2, 1 for programmed and 0 for empty.
.TP
\fB\-a\fR
get all the information above.
.TP
\fB\-c\fR
get YubiKey capability information.
\fB\-q\fR
modifier, only show the relevant data from the YubiKey, no extra information.
.TP
\fB\-V\fR
print tool version and exit

.SH EXAMPLE
Output of all information from a YubiKey

.nf
$ \fBykinfo \-s \-m \-H \-v \-t \-p\fR
serial: 1077254
serial_hex: 107006
serial_modhex: bcicch
version: 2.2.3
touch_level: 1793
programming_sequence: 1
$
.fi

.SH BUGS
Report ykinfo bugs in
.URL "https://github.com/Yubico/yubikey-personalization/issues" "the issue tracker"
.SH "SEE ALSO"
The
.URL "https://developers.yubico.com/yubikey-personalization/" "ykpersonalize home page"
.PP
YubiKeys can be obtained from
.URL "http://www.yubico.com/" "Yubico" "."


NEWS
Yubikey-personalize NEWS -- History of user-visible changes.     -*- outline -*-

* Version 1.17.1 (released 2015-04-01)

** Fixup of 1.17.0

* Version 1.17.0 (released 2015-04-01)

** add yk_get_capabilities() to fetch capabilities.

** add -c to ykinfo to fetch capabilities.

** whitelist firmware 4.1.x

* Version 1.16.4 (released 2015-03-23)

** change the tool to accept autoeject time as a short instead of a byte

* Version 1.16.3 (released 2015-03-10)

** whitelist YubiKey version 3.4.x

** only try to set libusb configuration if it's unset on the device

* Version 1.16.2 (released 2014-11-28)

** ykinfo: fix modhex printout when serial is an odd number of hex digits.

** whitelist yubikey version 4.0.x

** try to open more PIDs and add for udev.

* Version 1.16.1 (released 2014-10-30)

** Fix udev rules so they contain four digits.

** Only try to detach the kernel driver if it's attached.
For libusb-1.0

** Let import config report errors properly.

* Version 1.16.0 (released 2014-09-26)

** Add yk_get_key_vid_pid() to get the vendor and product id of a key.

** Add flags for ykinfo to print vendor and product id.

** Fix a bug in the osx backend where it would return an error opening
a composite device with two hid interfaces.

* Version 1.15.3 (released 2014-09-04)

** Fix URLs for opensource.y.com -> developers.y.com move.

** Whitelist firmware version 3.3 and detect new PIDs.

* Version 1.15.2 (released 2014-07-30)

** Whitelist firmware version 2.5

** Read key when importing configuration.

** Fix formatting error in information about what is written to key.

** Check return codes when doinf NDEF writes.

* Version 1.15.1 (released 2014-03-12)

** Add -6 and -8 flags to ykchalresp to output the response as a truncated
OATH code. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Add -t to ykchalresp to use current time / 30 as challenge in TOTP
fashion. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Fix for a crash in the libusb-0.1 backend.
Thanks to Antti Keränen <detegr@gmail.com>

* Version 1.15.0 (released 2014-01-10)

** add new Neo version 3.2 as supported.

** let ykp_AES_key_from_password() return error if it can't get a random
salt.

** let -a without option mean ask for raw key on STDIN.

** let no -a given mean generate a random key

** add ykp_AES_key_from_raw() and ykp_HMAC_key_from_raw() that sets the
key from a byte array.

** remove -osalt= and possibility to generate key from password.

** add YK_ENODATA that's returned when no data is returned from
the device and no other underlying error is set.

* Version 1.14.2 (released 2013-12-30)

** let -m take challenge-response timeout and autoeject timeout
supplied with : beteen, so mode:cr-timeout:autoeject-timeout

** make sure to always wait 1 millisecond between writes to the key
newer platforms triggered a race where writes to the key could
happen in a very short (about 0.4 milliseconds) timespan and would
not be processed.

** minor fixes for the printing format.

* Version 1.14.1 (released 2013-09-16)

** Fix breakage with latest automake.

* Version 1.14.0 (released 2013-07-04)

** Fixups of import/export.
Add targetConfig to show in which slot a configuration is intended.
Possible memory leaks on error conditions.

** Add -d switch to ykpersonalize for dry-run.

** Add ykp_clear_config() for clearing configuration flags.

** Add getter functions for all configuration flags.

** Add -V to all tools to output version.

** Add ykp_get_acccode_type() and ykp_set_acccode_type()
Only to do with export, showing where the access code came from
in the ycfg.

** Add -1 and -2 options to ykinfo to show programming state.

* Version 1.13.0 (released 2013-04-22)

** Added import/export of ycfg-json format.
Invoke with -fjson to -s or -i
Add exported functions ykp_export_config() and ykp_import_config()

** Fixup output of flags when using ykp_write_config()

** Add binary builds for mac.

** Minor cleanups noticed during debian packaging.

* Version 1.12.0 (released 2013-03-14)

** Recognize firmwares 2.4 and 3.1.

** Add support for setting the new extflag LED_INV
When set the behaviour of the led on the YubiKey is inversed.

* Version 1.11.3 (released 2013-01-09)

** Fixup of broken release.

* Version 1.11.2 (released 2013-01-09)

** Fix a bug where writing a NDEF with unknown prefix ended up writing invalid
data to the YubiKey NEO. Wrote prefix as 0x24 instead of 0x00.

** Don't allow opening a YubiKey if there's more than one present in the system.

** Fix shared linking of ykinfo and ykchalresp.

* Version 1.11.1 (released 2012-12-21)

** Implement ykusb_strerror() on windows.

** Fix a bug where a YubiKey would fail to be recognized if there was
another device from Yubico (vendor id 1050) inserted and looked at
before in the device chain.

** Fix a bug where you could only set 8 bytes of the public id with
the command line tool, now all 16 bytes can be set.

** Documentation updates and fixes.

* Version 1.11.0 (released 2012-12-12)

** Added version symbols and functions.
The header file is "ykpers-version.h" and it contains the following
symbols and functions: YKPERS_VERSION_STRING, YKPERS_VERSION_NUMBER,
YKPERS_VERSION_MAJOR, YKPERS_VERSION_MINOR, YKPERS_VERSION_PATCH,
ykpers_check_version.  Requested by Giampaolo Bellini
<iw2lsi@gmail.com>.

* Version 1.10.0 (released 2012-12-11)

** Support for the new productId of the production Neo.
Has ProducId 0x110, 0x111 or 0x112 depending on mode (see the notes about
-m and device_config).

** Add support for SLOT_NDEF2.
Use SLOT_NDEF to emit slot 1 as NDEF or SLOT_NDEF2 to emit slot 2.
This also adds the function yk_write_ndef2() that takes a slot parameter.

** Add -m flag for ykpersonalize, set usb mode of YubiKey NEO.
0 means pure YubiKey mode, 1 means pure CCID mode and 2 means YubiKey/CCID
composite mode. Add 80 to set EJECT_FLAG.
To use this with the api, see the functions:

ykinfo.c
/* -*- mode:C; c-file-style: "bsd" -*- */
/*
 * Copyright (c) 2012-2015 Yubico AB.
 * All rights reserved.
 *
 * Some basic code copied from ykchalresp.c.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <unistd.h>
#include <stdbool.h>

#include <yubikey.h>
#include <ykcore.h>
#include <ykstatus.h>
#include <ykpers-version.h>
#include <ykdef.h>

const char *usage =
	"Usage: ykinfo [options]\n"
	"\n"
	"Options :\n"
	"\n"
	"\t-s        Get serial in decimal from YubiKey\n"
	"\t-m        Get serial in modhex from YubiKey\n"
	"\t-H        Get serial in hex from YubiKey\n"
	"\t-v        Get version from YubiKey\n"
	"\t-t        Get touchlevel from YubiKey\n"
	"\t-1        Check if slot 1 is programmed\n"
	"\t-2        Check if slot 2 is programmed\n"
	"\t-p        Get programming sequence from YubiKey\n"
	"\t-i        Get vendor id of YubiKey\n"
	"\t-I        Get product id of YubiKey\n"
	"\t-a        Get all information above\n"
	"\t-c        Get capabilities from YubiKey\n"
	"\n"
	"\t-q        Only output information from YubiKey\n"
	"\n"
	"\t-V        Get the tool version\n"
	"\t-h        help (this text)\n"
	"\n"
	"\n"
	;
const char *optstring = "asmHvtpqhV12iIc";

static void report_yk_error(void)
{
	if (yk_errno) {
		if (yk_errno == YK_EUSBERR) {
			fprintf(stderr, "USB error: %s\n",
				yk_usb_strerror());
		} else {
			fprintf(stderr, "Yubikey core error: %s\n",
				yk_strerror(yk_errno));
		}
	}
}

static int parse_args(int argc, char **argv,
		bool *serial_dec, bool *serial_modhex, bool *serial_hex,
		bool *version, bool *touch_level, bool *pgm_seq, bool *quiet,
		bool *slot1, bool *slot2, bool *vid, bool *pid, bool *capa,
		int *exit_code)
{
	int c;

	while((c = getopt(argc, argv, optstring)) != -1) {
		switch (c) {
		case 'a':
			*serial_dec = true;
			*serial_modhex = true;
			*serial_hex = true;
			*version = true;
			*touch_level = true;
			*pgm_seq = true;
			*slot1 = true;
			*slot2 = true;
			*vid = true;
			*pid = true;
			break;
		case 's':
			*serial_dec = true;
			break;
		case 'm':
			*serial_modhex = true;
			break;
		case 'H':
			*serial_hex = true;
			break;
		case 'v':
			*version = true;
			break;
		case 't':
			*touch_level = true;
			break;
		case 'p':
			*pgm_seq = true;
			break;
		case 'q':
			*quiet = true;
			break;
		case '1':
			*slot1 = true;
			break;
		case '2':
			*slot2 = true;
			break;
		case 'i':
			*vid = true;
			break;
		case 'I':
			*pid = true;
			break;
		case 'c':
			*capa = true;
			break;
		case 'V':
			fputs(YKPERS_VERSION_STRING "\n", stderr);
			*exit_code = 0;
			return 0;
		case 'h':
		default:
			fputs(usage, stderr);
			*exit_code = 0;
			return 0;
		}
	}

	if (!*serial_dec && !*serial_modhex && !*serial_hex &&
			!*version && !*touch_level && !*pgm_seq && !*slot1 && !*slot2 &&
			!*vid && !*pid && !*capa) {
		/* no options at all */
		fputs("You must give at least one option.\n", stderr);
		fputs(usage, stderr);
		return 0;
	}

	return 1;
}


int main(int argc, char **argv)
{
	YK_KEY *yk = 0;
	bool error = true;
	int exit_code = 0;

	/* Options */
	bool serial_dec = false;
	bool serial_modhex = false;
	bool serial_hex = false;
	bool version = false;
	bool touch_level = false;
	bool pgm_seq = false;
	bool slot1 = false;
	bool slot2 = false;
	bool vid = false;
	bool pid = false;
	bool capa = false;

	bool quiet = false;

	yk_errno = 0;

	if (! parse_args(argc, argv,
				&serial_dec, &serial_modhex, &serial_hex,
				&version, &touch_level, &pgm_seq, &quiet,
				&slot1, &slot2, &vid, &pid, &capa,
				&exit_code))
		exit(exit_code);

	if (!yk_init()) {
		exit_code = 1;
		goto err;
	}

	if (!(yk = yk_open_first_key())) {
		exit_code = 1;
		goto err;
	}

	if(serial_dec || serial_modhex || serial_hex) {
		unsigned int serial;
		int ret = yk_get_serial(yk, 1, 0, &serial);
		if(!ret) {
			exit_code = 1;
			goto err;
		}
		if(serial_dec) {
			if(!quiet)
				printf("serial: ");
			printf("%d\n", serial);
		}
		if(serial_modhex || serial_hex) {
			char buf[64];
			char hex_serial[64];
			char modhex_serial[64];
			char *ptr = buf;

			int chars = snprintf(buf + 1, 63, "%x", serial);
			if(chars % 2 == 1) {
				buf[0] = '0';
			} else {
				ptr += 1;
			}
			if(serial_hex) {
				if(!quiet)
					printf("serial_hex: ");
				printf("%s\n", ptr);
			}
			if(serial_modhex) {
				yubikey_hex_decode(hex_serial, ptr, strlen(ptr));
				yubikey_modhex_encode(modhex_serial, hex_serial, strlen(hex_serial));
				if(!quiet)
					printf("serial_modhex: ");
				printf("%s\n", modhex_serial);
			}
		}
	}
	if(version || touch_level || pgm_seq || slot1 || slot2) {
		YK_STATUS *st = ykds_alloc();
		if(!yk_get_status(yk, st)) {
			ykds_free(st);
			exit_code = 1;
			goto err;
		}

		if(version) {
			if(!quiet)
				printf("version: ");
			printf("%d.%d.%d\n", ykds_version_major(st), ykds_version_minor(st), ykds_version_build(st));
		}
		if(touch_level) {
			if(!quiet)
				printf("touch_level: ");
			printf("%d\n", ykds_touch_level(st));
		}
		if(pgm_seq) {
			if(!quiet)
				printf("programming_sequence: ");
			printf("%d\n", ykds_pgm_seq(st));
		}
		if(slot1) {
			if(!quiet)
				printf("slot1_status: ");
			printf("%d\n", (ykds_touch_level(st) & CONFIG1_VALID) == CONFIG1_VALID);
		}
		if(slot2) {
			if(!quiet)
				printf("slot2_status: ");
			printf("%d\n", (ykds_touch_level(st) & CONFIG2_VALID) == CONFIG2_VALID);
		}
		ykds_free(st);
	}
	if(vid || pid) {
		int vendor_id, product_id;
		if(!yk_get_key_vid_pid(yk, &vendor_id, &product_id)) {
			exit_code = 1;
			goto err;
		}
		if(vid) {
			if(!quiet)
				printf("vendor_id: ");
			printf("%x\n", vendor_id);
		}
		if(pid) {
			if(!quiet)
				printf("product_id: ");
			printf("%x\n", product_id);
		}
	}
	if(capa) {
		unsigned char buf[0xff];
		unsigned int len = 0xff;
		unsigned int i;
		if(!yk_get_capabilities(yk, 1, 0, buf, &len)) {
			exit_code = 1;
			goto err;
		}
		if(!quiet)
			printf("capabilities: ");
		for(i = 0; i < len; i++) {
			printf("%02x", buf[i]);
		}
		printf("\n");
	}

	exit_code = 0;
	error = false;

err:
	if (error || exit_code != 0) {
		report_yk_error();
	}

	if (yk && !yk_close_key(yk)) {
		report_yk_error();
		exit_code = 2;
	}

	if (!yk_release()) {
		report_yk_error();
		exit_code = 2;
	}

	exit(exit_code);
}
